### ISSUE TITLE
standard UTF8 encoding for MediaWiki databases

### ISSUE THREAD

**Issue by Tigerfell** [openstreetmap/operations#373]
I would like to ask you to change the character encoding of the MySQL databases for MediaWiki installations to `utf8mb4` encoding. It looks like they currently use utf8 which means that it uses three Bytes to store a character. It is not a real UTF8 encoding, many characters can not be stored. This results in wiki pages being trimmed when someone enters a non-supported character [1]. There are currently two use cases which require "standard" UTF8 encoding.

- A user wants to write their OSM user name in the wiki. [1]
- A MediaWiki gadget is used and the internal UTF8 test fails, which renders it useless. [2]

Additionally, the current encoding is deprecated according to MySQL 8 documentation [3] and will be removed. The wiki currently uses MySQL 5.7.29.

[1] https://wiki.openstreetmap.org/w/index.php?title=Bot&diff=prev&oldid=1784135
[2] https://wiki.openstreetmap.org/wiki/MediaWiki:Gadget-HotCat.js
[3] https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-sets.html

**Comment by tomhughes** [comment]
As I said on email a week ago, doing the conversion is easy enough though it will need about half an hour of downtime for the main wiki.

My big concern is how to make sure that any new schema changes made by future mediawiki updates use the new encoding - the current tables do not appear to be using the default database
encoding which suggests that mediawiki is setting the encoding explicitly when it creates tables and columns.

**Comment by matthewdarwin** [comment]
Set the the default in the server settings config file?
```
default-character-set = utf8mb4
```

**Comment by tomhughes** [comment]
Which will achieve what exactly?

I've already explained that the existing tables have a character set that does not match the current default character set, so I have to assume that mediawiki specified a character set explicitly when creating them, and would do so again for any new columns or tables.

Changing the default would do nothing at all to change that.

**Comment by tomhughes** [comment]
Looks like mediawiki actually reject using utf8mb4 (https://phabricator.wikimedia.org/T50767) and actually reccomend using the binary encoding so that mysql just preserves whatever bytes mediawiki throws at it.

That is of course an insane solution, so probably what you'd expect from mediawiki :-(

**Comment by tomhughes** [comment]
If I understand https://phabricator.wikimedia.org/T196092 correctly then traditionally mediawiki had an option to enabled UTF-8 during installation which would cause to explicitly use the utf8 encoding for tables but that have now removed that option.

I just need to figure out how that impacts upgrades to existing installations and how to change it if necessary...

**Comment by tomhughes** [comment]
See also https://phabricator.wikimedia.org/T194125 where they seem to suggest that utf8mb4 won't actually work. That was closed with a reference to https://phabricator.wikimedia.org/T191231 as the solution and that is still open :-(

**Comment by 1ec5** [comment]
> This results in wiki pages being trimmed when someone enters a non-supported character

As a workaround, the user can invoke [this Scribunto module](https://wiki.openstreetmap.org/wiki/Module:Char) with Unicode codepoints.

(So that others can find this issue more easily: this issue tracks the inability to insert literal emoji and certain CJK characters, among other things.)

**Comment by pnorman** [comment]
This will likely require dumping the wiki and reloading it and needs a priority set.

**Comment by Firefishy** [comment]
> This will likely require dumping the wiki and reloading it and needs a priority set.

Mediawiki does not have an internal backup method which backups all users and content. 
The [documentation on mediawiki.org backup method is to dump a copy of the database](https://www.mediawiki.org/wiki/Manual:Backing_up_a_wiki) which would not fix any encoding issues. [DumpBackup.php]

(https://www.mediawiki.org/wiki/Manual:DumpBackup.php) quote: "_XML dumps contain the content of the wiki (wiki pages with all their revisions), without the site-related data. DumpBackup.php does not create a full backup of the wiki database, the dump does not contain user accounts, images, edit logs, deleted revisions, etc._"

**Comment by Firefishy** [comment]
This ticket has no concrete description of what action is needed. Please describe the steps required or point to documentation which shows what needs fixing.

The relevant [LocalSettings.php](https://github.com/openstreetmap/chef/blob/3e9671df3fe0b4ff0353f02adc326b3f9b80a33e/cookbooks/mediawiki/templates/default/LocalSettings.php.erb#L76-L80) are described here. We have it set to `binary`

**Comment by Firefishy** [comment]
The closest thing I can find to "document" the change required is [this throw away comment](https://phabricator.wikimedia.org/T194125#5331955) about running quote: _To migrate to binary run: `ALTER TABLE table CONVERT TO CHARACTER SET binary;` on all your tables_`

I need proper documentation.

**Comment by tomhughes** [comment]
That sounds very wrong - isn't this about moving from mysql's weird `utf8` to the more standard (but oddly named) `utf8mb4`?

**Comment by Firefishy** [comment]
> That sounds very wrong - isn't this about moving from mysql's weird `utf8` to the more standard (but oddly named) `utf8mb4`?

My understanding is they gave up on getting the extended unicode working with MySQL's `utf8mb4` etc and went with an internal implementation in PHP instead, while using `binary` in MySQL. There appears to be no proper documentation on "fixing" an existing database and I am not going to haphazardly proceed without.

**Comment by 1ec5** [comment]
Correct, here’s a MediaWiki developer explaining [some of the history](https://bikeshed.vibber.net/@brion/111106533208904485) behind this unusually complex implementation.

**Comment by Firefishy** [comment]
I would absolutely love to fix this issue, but I need some more clear guidance on what the issue is and what the fix might be. If any data is needed I can extract it.

**Comment by tomhughes** [comment]
I've run a conversion on https://test.wiki.openstreetmap.org/ to binary which I believe is the mediawiki recommendation even if that is mad.

One exception is `searchindex` which I converted to utf8mb4 instead as that is what a clean install appears to do.

It certainly resolves the emoji problem from #1235 so if nobody can see any problems then we could run the conversion in production - it only took about 15 minutes or so I think.

**Comment by Hufkratzer** [comment]
> I would absolutely love to fix this issue, but I need some more clear guidance on what the issue is and what the fix might be. If any data is needed I can extract it.

I am the one who caused all this trouble by adding a new topic to this wiki page 3 weeks ago that contained some emojis. Sorry! The preview worked. When I hit the save button, I got a warning like this: 
![Image](https://github.com/user-attachments/assets/6d22fe54-2e35-49ec-a6ab-9b4a2068d454)

So I replaced the emojis as it was recomended in the warning. After saving the page still worked well, until recently.

I could remove the emojis and/or colons from the text if I could edit the page, but that doesn't seem to be possible.

EDIT / ADDITION

I was wrong; I could edit the page and removed the emojis.

**Comment by 1ec5** [comment]
> It certainly resolves the emoji problem from https://github.com/openstreetmap/operations/issues/1235

Are you sure? I’m still seeing the same error viewing [this old revision](https://wiki.openstreetmap.org/wiki/Special:PermanentLink/2837089) from before @Hufkratzer [removed the emojis](https://wiki.openstreetmap.org/wiki/Special:Diff/2848763) from the page.

**Comment by 1ec5** [comment]
On the test wiki, interwiki prefixes no longer resolve as external links, for example in [this article](https://test.wiki.openstreetmap.org/wiki/Sidewalks).

**Comment by tomhughes** [comment]
> Are you sure? I’m still seeing the same error viewing [this old revision](https://wiki.openstreetmap.org/wiki/Special:PermanentLink/2837089) from before [@Hufkratzer](https://github.com/Hufkratzer) [removed the emojis](https://wiki.openstreetmap.org/wiki/Special:Diff/2848763) from the page.

It's the test wiki I fixed, not the production one.

**Comment by 1ec5** [comment]
Ah, my bad, I [applied one of the headings](https://test.wiki.openstreetmap.org/wiki/Special:Diff/2826000) and it does work. 👍

**Comment by tomhughes** [comment]
The encoding conversion had apparently nul padded the strings in the interwiki table.

I've fixed it now but I will need to do more investigation into the wider effects of the conversion.

**Comment by 1ec5** [comment]
Posted a call for testers to [the wiki](https://wiki.openstreetmap.org/wiki/Talk:Wiki#Unicode_conversion_ready_for_testing) and [the forum](https://community.openstreetmap.org/t/help-test-a-fix-for-the-wiki-s-unicode-support/130732).

**Comment by 1ec5** [comment]
> The encoding conversion had apparently nul padded the strings in the interwiki table.

We have [a report](https://community.openstreetmap.org/t/help-test-a-fix-for-the-wiki-s-unicode-support/130732/6) that the null padding still affects searching for an interwiki prefix.

**Comment by 1ec5** [comment]
Additionally, page title prefixes like “Hr:” are [redirecting to Wikipedia articles](https://community.openstreetmap.org/t/help-test-a-fix-for-the-wiki-s-unicode-support/130732/7) instead of displaying pseudonamespaced pages on our wiki. Somehow the test wiki has [all the interwiki prefixes](https://test.wiki.openstreetmap.org/wiki/Special:Interwiki#interwikitable-language) installed, unlike the production wiki. Each of these prefixes has [95 U+FFFF characters](https://test.wiki.openstreetmap.org/w/index.php?title=Special:Interwiki&action=edit&prefix=aa) appended to the end of the URL.

**Comment by tomhughes** [comment]
I the extra interwiki prefixes is probably my fault - when I was trying to resolve the issue that turned out to be down to nul padding I think I ran a maintenace script that repopulated the table... I'll see if I can take a new copy of the table from the production instance.

**Comment by tomhughes** [comment]
Hopefully that should be better now.

**Comment by 1ec5** [comment]
Yep, it’s a bit better, though I think you also blew away the `w:` prefix for the English Wikipedia. It’s fine as long as we aren’t planning to synchronize the interwiki table back to the production site.

**Comment by tomhughes** [comment]
I suspect that was already broken? The issue is that the defined prefix is `W` but previously the table had a case insensitive collation and now it doesn't, but I kept the existing post conversion schema and just restored the data.

Looking at our other wikis which were installed more recently and use binary encoding they don't have any collation set either.

**Comment by 1ec5** [comment]
Oh, maybe so. That reminds me of another longstanding issue in Wikibase, possibly also related to collation settings: #1250.

**Comment by tomhughes** [comment]
So in fact it seems mediawiki always lowercases the prefix before looking it up, but for whatever reason we had some prefixes in uppercase which worked because of the case insensitive collation.

I've converted all prefixes in the test wiki to lowercase now and both `W` and `w` now seem to work.

**Comment by 1ec5** [comment]
The only feedback we got from [the call I put out on the forum in May](https://community.openstreetmap.org/t/help-test-a-fix-for-the-wiki-s-unicode-support/130732) was about the broken interwiki prefixes, which were fixed in https://github.com/openstreetmap/operations/issues/373#issuecomment-2953828746. Do we know of anything else that we’d be concerned about before bringing the changes to production? If we don’t feel confident about it yet, I can put out another call for testing.

**Comment by tomhughes** [comment]
The conversion has now been run on the main wiki.

**Comment by tomhughes** [comment]
For the record this is the set of commands I ran to do the main conversion:

```sql
alter table abuse_filter convert to character set binary, character set = binary;
alter table abuse_filter_action convert to character set binary, character set = binary;
alter table abuse_filter_history convert to character set binary, character set = binary;
alter table abuse_filter_log convert to character set binary, character set = binary;
alter table archive convert to character set binary, character set = binary;
alter table blobs convert to character set binary, character set = binary;
alter table brokenlinks convert to character set binary, character set = binary;
alter table category convert to character set binary, character set = binary;
alter table categorylinks convert to character set binary, character set = binary;
alter table change_tag convert to character set binary, character set = binary;
alter table cu_changes convert to character set binary, character set = binary;
alter table cu_log convert to character set binary, character set = binary;
alter table cur convert to character set binary, character set = binary;
alter table external_user convert to character set binary, character set = binary;
alter table externallinks convert to character set binary, character set = binary;
alter table filearchive convert to character set binary, character set = binary;
alter table image convert to character set binary, character set = binary;
alter table imagelinks convert to character set binary, character set = binary;
alter table interwiki convert to character set binary, character set = binary;
alter table ipblocks_old convert to character set binary, character set = binary;
alter table iwlinks convert to character set binary, character set = binary;
alter table job convert to character set binary, character set = binary;
alter table l10n_cache convert to character set binary, character set = binary;
alter table langlinks convert to character set binary, character set = binary;
alter table links convert to character set binary, character set = binary;
alter table linkscc convert to character set binary, character set = binary;
alter table log_search convert to character set binary, character set = binary;
alter table logging convert to character set binary, character set = binary;
alter table math convert to character set binary, character set = binary;
alter table module_deps convert to character set binary, character set = binary;
alter table objectcache convert to character set binary, character set = binary;
alter table oldimage convert to character set binary, character set = binary;
alter table page convert to character set binary, character set = binary;
alter table page_props convert to character set binary, character set = binary;
alter table page_restrictions convert to character set binary, character set = binary;
alter table pagelinks convert to character set binary, character set = binary;
alter table protected_titles convert to character set binary, character set = binary;
alter table querycache convert to character set binary, character set = binary;
alter table querycache_info convert to character set binary, character set = binary;
alter table querycachetwo convert to character set binary, character set = binary;
alter table recentchanges convert to character set binary, character set = binary;
alter table redirect convert to character set binary, character set = binary;
alter table revision convert to character set binary, character set = binary;
alter table site_stats convert to character set binary, character set = binary;
alter table spoofuser convert to character set binary, character set = binary;
alter table templatelinks convert to character set binary, character set = binary;
alter table text convert to character set binary, character set = binary;
alter table trackbacks convert to character set binary, character set = binary;
alter table updatelog convert to character set binary, character set = binary;
alter table uploadstash convert to character set binary, character set = binary;
alter table user convert to character set binary, character set = binary;
alter table user_former_groups convert to character set binary, character set = binary;
alter table user_groups convert to character set binary, character set = binary;
alter table user_newtalk convert to character set binary, character set = binary;
alter table user_properties convert to character set binary, character set = binary;
alter table user_rights convert to character set binary, character set = binary;
alter table validate convert to character set binary, character set = binary;
alter table watchlist convert to character set binary, character set = binary;
```

Not all tables are included there because quite a few (probably ones added more recently) were already using the binary encoding.

I then fixed up the `interwiki` table which had a schema that was for some reason very different to a new install, and normalised the case of the prefixes, with:

```sql
alter table interwiki modify column iw_prefix varbinary(32) not null;
alter table interwiki modify column iw_url blob;
alter table interwiki alter column iw_local drop default;
alter table interwiki modify column iw_trans tinyint(4) not null default 0;
update interwiki set iw_prefix = trim(trailing '\0' from iw_prefix);
update interwiki set iw_prefix = lower(convert(iw_prefix using utf8mb4));
```

**Comment by 1ec5** [comment]
Following the conversion, I’ve disabled [the abuse filter](https://wiki.openstreetmap.org/wiki/Special:AbuseFilter/13) that had been stopping users from entering literal emoji and such.
